### PR TITLE
Added multi-agent support to R2P2

### PR DESCRIPTION
**Pull request by alvkao58**
Extends R2P2 to the multi-agent case, outputting predictions for all vehicles within a specified radius. 

For each vehicle, the operator rotates the ego-vehicle's LIDAR point cloud, as well as the vehicle's past trajectory, to that vehicle's coordinate frame before performing inference. Orientations of vehicles are estimated using the past two locations of the vehicle. R2P2 inherently does not perform any joint modeling of agents.

### CODE REVIEW

File: pylot/prediction/flags.py [hunk 1]
```diff
@@ -6,6 +6,12 @@
 flags.DEFINE_integer(
     'prediction_num_future_steps', None,
     'Number of future steps outputted by the prediction module.')
+flags.DEFINE_integer(
+    'prediction_radius', 50,
+    'Make predictions for all vehicles within this radius of the ego-vehicle.')
+flags.DEFINE_boolean(
```
Comment (ICGog) [inline pylot/prediction/flags.py:12]: Can we please have this False by default? I think the planners break if we send predictions for ego. We can adjust them in another PR.

File: pylot/prediction/r2p2_predictor_operator.py [hunk 2]
```diff
@@ -38,150 +37,253 @@ class R2P2PredictorOperator(erdos.Operator):
...
-        num_past_locations = ego_vehicle_trajectory.shape[0]
-        if num_past_locations < self._flags.prediction_num_past_steps:
-            initial_copies = np.repeat([np.array(ego_vehicle_trajectory[0])],
-                                       self._flags.prediction_num_past_steps - num_past_locations,
-                                       axis=0)
-            ego_vehicle_trajectory = np.vstack(
-                (initial_copies, ego_vehicle_trajectory))
-        ego_vehicle_trajectory = np.expand_dims(ego_vehicle_trajectory, axis=0)
-
+        all_vehicles = [trajectory for trajectory in tracking_msg.obstacle_trajectories if trajectory.label == 'vehicle']
```
Comment (ICGog) [inline pylot/prediction/r2p2_predictor_operator.py:80]: Can you please check if label is in pylot.perception.detection.utils.VEHICLE_LABELS? This makes sure that the code works even if we use real perception.

File: pylot/prediction/r2p2_predictor_operator.py [hunk 3]
```diff
@@ -38,150 +37,253 @@ class R2P2PredictorOperator(erdos.Operator):
...
+        closest_vehicles, ego_vehicle = self._get_closest_vehicles(all_vehicles)
+        self._logger.info('Getting predictions for {} vehicles'.format(
```
Comment (ICGog) [inline pylot/prediction/r2p2_predictor_operator.py:82]: Please timestamp the log event.

File: pylot/prediction/r2p2_predictor_operator.py [hunk 4]
```diff
@@ -38,150 +37,253 @@ class R2P2PredictorOperator(erdos.Operator):
...
+            len(closest_vehicles)))
+
+        start_time = time.time()
+        if len(closest_vehicles) == 0:
+            self._logger.debug('@{}: no vehicles to make predictions for'.format(timestamp))
+            prediction_stream.send(PredictionMessage(timestamp, []))
+            return
+        closest_vehicles_ego_transforms = self._get_closest_vehicles_ego_transforms(closest_vehicles, ego_vehicle)
+
+        num_predictions = len(closest_vehicles)
```
Comment (ICGog) [inline pylot/prediction/r2p2_predictor_operator.py:92]: Please move it several lines up, and use it the code above.

File: pylot/prediction/r2p2_predictor_operator.py [hunk 5]
```diff
@@ -38,150 +37,253 @@ class R2P2PredictorOperator(erdos.Operator):
...
+                    location=Location(cur_point[0], cur_point[1], cur_point[2]),
+                    rotation=closest_vehicles_ego_transforms[idx].rotation))
+ 
+            obstacle_predictions_list.append(
+                ObstaclePrediction('vehicle',
+                                   closest_vehicles[idx].id,
+                                   closest_vehicles_ego_transforms[idx],
+                                   closest_vehicles[idx].bounding_box,
+                                   1.0, # Probability; currently a filler value because we are taking just one sample from distribution
+                                   predictions))
```
Comment (ICGog) [inline pylot/prediction/r2p2_predictor_operator.py:164]: Can you please move this code into a separate method?

File: pylot/prediction/r2p2_predictor_operator.py [hunk 6]
```diff
@@ -38,150 +37,253 @@ class R2P2PredictorOperator(erdos.Operator):
...
+                                       axis=0)
+            trajectory = np.vstack(
+                (initial_copies, trajectory))
+        elif num_past_locations > self._flags.prediction_num_past_steps:
+            trajectory = trajectory[-self._flags.prediction_num_past_steps:]
+        assert trajectory.shape[0] == self._flags.prediction_num_past_steps
+        return trajectory
+
+
+
```
Comment (ICGog) [inline pylot/prediction/r2p2_predictor_operator.py:192]: Superfluous empty lines.

File: pylot/prediction/r2p2_predictor_operator.py [hunk 7]
```diff
@@ -38,150 +37,253 @@ class R2P2PredictorOperator(erdos.Operator):
...
+        sorted_vehicles = [v for v, d in sorted(
+            zip(all_vehicles, distances), key=lambda pair: pair[1])
+            if d <= self._flags.prediction_radius]
+
+        if self._flags.prediction_ego_agent:
+            return sorted_vehicles, sorted_vehicles[0]
+        else: # Exclude the ego vehicle.
+            return sorted_vehicles[1:], sorted_vehicles[0]
+
+
```
Comment (ICGog) [inline pylot/prediction/r2p2_predictor_operator.py:206]: Superfluous empty line.

File: pylot/prediction/r2p2_predictor_operator.py [hunk 8]
```diff
@@ -38,150 +37,253 @@ class R2P2PredictorOperator(erdos.Operator):
...
+        for i in range(len(closest_vehicles)):
+            cur_vehicle_angle = self._get_vehicle_orientation(closest_vehicles[i])
+            closest_vehicles_ego_transforms.append(
+                Transform(location=closest_vehicles_ego_locations[i].location, 
+                          rotation=Rotation(yaw=cur_vehicle_angle)
+            ))
+        return closest_vehicles_ego_transforms
+
+
+    def _get_vehicle_orientation(self, vehicle):
```
Comment (ICGog) [inline pylot/prediction/r2p2_predictor_operator.py:223]: Shouldn't this be a method of DetectedObstacle class?

Comment (alvkao58) [reply pylot/prediction/r2p2_predictor_operator.py:224]: `_get_vehicle_orientation` uses the past locations of the vehicle to estimate its heading, so I'm not sure it fits in the DetectedObstacle class.

**Pull request merged**